genetIC/src/simulation/modifications/linearmodification.hpp

```diff
@@ -257,37 +246,167 @@ namespace modifications {
+      // multi-level Poisson solution. Once Fourier modes are combined on zoom grids, there is no simple way to get the potential.
+      OverdensityModification<DataType,T>::turnLocalisationCovectorIntoModificationCovector(field);
+      this->computeVelocity(field, this->direction);
+    }
+  };
+
+  template<typename DataType, typename T=tools::datatypes::strip_complex<DataType>>
+  class AngMomentumModification : public OverdensityModification<DataType, T> {
+  protected:
+    int direction;
+
+  /*! \brief Construct an angular momentum modification from a specified multi-level context, cosmology, and velocity component
+
+      \param underlying_ - underlying multi-level context object.
+      \param cosmology_ - struct containing cosmological data
+      \param direction - component of t velocity to modify, (0,1,2) <-> (x,y,z).
+
+      This evaluates the integral \int_\Gamma (q-\bar{q}) x (d Psi / dt) d³q,
+      where q is the Lagrangian position, \bar{q} is the mean withing the Lagrangian
+      patch and Psi is the rescaled potential. Note that this integral is done in
+      Lagrangian space (d³q) over the Lagrangian patch.
+  */
+
+  public:
+    AngMomentumModification(const multilevelgrid::MultiLevelGrid<DataType> &underlying_,
+                            const cosmology::CosmologicalParameters<T> &cosmology_, int direction_) :
+        OverdensityModification<DataType, T>(underlying_, cosmology_), direction(direction_) {
+      if (direction_ < 0 || direction_ > 2)
+        throw std::runtime_error("Angular momentum direction must be 0 (x), 1 (y) or 2 (z)");
+
+#ifdef ZELDOVICH_GRADIENT_FOURIER_SPACE
+      throw std::runtime_error("Refusing to continue as the code was compiled with the 'ZELDOVICH_GRADIENT_FOURIER_SPACE' activated, but an angular momentum modification was applied. These modifications require the gradient to be performed using a finite-difference scheme ; otherwise the output may not satisfy the modifications. Comment out this error if you are sure about what you are doing.");
+#endif
+      direction = direction_;
+    };
+
+  protected:
+    fields::Field<DataType, T> calculateLocalisationCovector(const grids::Grid<T> &grid) override {
+
+      auto qcenter = this->getCentre(grid);
+      T a = -1. / 12. / grid.cellSize, b = 2. / 3. / grid.cellSize;  // signs here so that L ~ - Nabla Phi
+
+      int dirp1 = (direction + 1) % 3,
+          dirp2 = (direction + 2) % 3;
+
+      fields::Field<DataType, T> outputField = fields::Field<DataType, T>(grid, false);
+      std::vector<DataType> &outputData = outputField.getDataVector();
+
+      for (size_t i = 0; i < this->flaggedCellsFinestGrid.size(); ++i) {
+        size_t index = this->flaggedCellsFinestGrid[i];
+        Coordinate<T> q = grid.getCentroidFromIndex(index);
+
+        Coordinate<T> deltaq = q - qcenter;
+
+        // Wrap around domain size
+        for (int dir = 0; dir < 3; ++dir) {
```

Comment:
I think this would be neater as `deltaq = grid.getWrappedOffset(qcenter,q)` if you agree that is equivalent?